File: LocalisationMateriel/src/java/physique/BatimentDataServiceImpl.java
package physique;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import metier.Batiment;
import metier.Salle;

public class BatimentDataServiceImpl implements BatimentDataService {

    private final ServiceSQL sql;
    private Connection cnx;
    private PreparedStatement statement;
    private ResultSet result;

    BatimentDataServiceImpl() {
        sql = new ServiceSQLImpl();
    }
    
    private void close() throws SQLException, Exception{
        if(!this.result.isClosed()){
            this.result.close();
        }
        if(!this.statement.isClosed()){
            this.statement.close();
        }
        if(!this.cnx.isClosed()){
            this.cnx.close();
        }
        this.sql.close();
    }

    private Batiment createBatiment(ResultSet result) throws Exception {
        Salle salle;
        long id = result.getLong("id");
        String nomBatiment = result.getString("nomBatiment");
        List<Salle> salles = new ArrayList<Salle>();
        String plan = result.getString("plan");
        String query = "SELECT * FROM salle WHERE id IN (SELECT idSalle FROM attribution WHERE idBatiment = ?) ORDER BY numSalle";
        PreparedStatement statementCreate = this.cnx.prepareStatement(query);
        statementCreate.setLong(1, id);
        ResultSet resultCreate = statementCreate.executeQuery();
        while (resultCreate.next()) {
            long idSalle = resultCreate.getLong("id");
            int x_ini = resultCreate.getInt("x_ini");
            int y_ini = resultCreate.getInt("y_ini");
            int x_fin = resultCreate.getInt("x_fin");
            int y_fin = resultCreate.getInt("y_fin");
            int numSalle = resultCreate.getInt("numSalle");
            String nomSalle = resultCreate.getString("nomSalle");
            String nomAtelier = resultCreate.getString("nomAtelier");
            salle = new Salle(x_ini, y_ini, x_fin, y_fin, numSalle, nomSalle, nomAtelier);
            salle.setId(idSalle);
            salles.add(salle);
        }
        Batiment batiment = new Batiment(nomBatiment, salles, plan);
        batiment.setId(id);
        if(!resultCreate.isClosed()){
            resultCreate.close();
        }
        if(!statementCreate.isClosed()){
            statementCreate.close();
        }
        return batiment;
    }

    @Override
    public Batiment add(Batiment batiment) throws Exception {
        cnx = this.sql.connect();
        Batiment newBatiment = null;
        boolean execute = false;
        long id = 0;
        List<Long> idSallesNonAttribue = new ArrayList<Long>();
        String query = "INSERT INTO batiment (nomBatiment, plan) VALUES (?, ?)";
        statement = this.cnx.prepareStatement(query, PreparedStatement.RETURN_GENERATED_KEYS);
        statement.setString(1, batiment.getNomBatiment());
        statement.setString(2, batiment.getPlan());
        if (statement.executeUpdate() > 0) {
            execute = true;
            result = statement.getGeneratedKeys();
            if (result.next()) {
                id = result.getLong(1);
            }
        }
        if (execute) {
            query = "SELECT id FROM salle WHERE id NOT IN (SELECT idSalle FROM attribution)";
            statement = this.cnx.prepareStatement(query);
            result = statement.executeQuery();
            while (result.next()) {
                idSallesNonAttribue.add(result.getLong("id"));
            }
            query = "INSERT INTO attribution (idBatiment, idSalle) VALUES (?, ?)";
            statement = this.cnx.prepareStatement(query);
            for (Salle salle : batiment.getSalles()) {
                if (idSallesNonAttribue.contains(salle.getId())) {
                    statement.setLong(1, id);
                    statement.setLong(2, salle.getId());
                    statement.executeUpdate();
                }
            }
        }
        if (id != 0) {
            newBatiment = getById(id);
        }
        this.close();
        return newBatiment;
    }

    @Override
    public boolean remove(Batiment batiment) throws Exception {
        cnx = this.sql.connect();
        boolean execute = false;
        String query = "DELETE FROM attribution WHERE idBatiment = ?";
        statement = this.cnx.prepareStatement(query);
        statement.setLong(1, batiment.getId());
        statement.executeUpdate();
        query = "DELETE FROM batiment WHERE id = ?";
        statement = this.cnx.prepareStatement(query);
        statement.setLong(1, batiment.getId());
        if (statement.executeUpdate() > 0) {
            execute = true;
        }
        this.close();
        return execute;
    }

    @Override
    public boolean update(Batiment batiment) throws Exception {
        cnx = this.sql.connect();
        List<Long> idSallesBatiment = new ArrayList<Long>();
        List<Long> idSallesNonAttribue = new ArrayList<Long>();
        boolean execute = false;
        String query = "UPDATE batiment SET nomBatiment = ?, plan = ? WHERE id = ?";
        statement = this.cnx.prepareStatement(query);
        statement.setString(1, batiment.getNomBatiment());
        statement.setString(2, batiment.getPlan());
        statement.setLong(3, batiment.getId());
        if (statement.executeUpdate() > 0) {
            execute = true;
        }
        if (execute) {
            query = "SELECT idSalle FROM attribution WHERE idBatiment = ?";
            statement = this.cnx.prepareStatement(query);
            statement.setLong(1, batiment.getId());
            result = statement.executeQuery();
            while (result.next()) {
                idSallesBatiment.add(result.getLong("idSalle"));
            }
            query = "SELECT id FROM salle WHERE id NOT IN (SELECT idSalle FROM attribution)";
            statement = this.cnx.prepareStatement(query);
            result = statement.executeQuery();
            while (result.next()) {
                idSallesNonAttribue.add(result.getLong("id"));
            }
            query = "INSERT INTO attribution (idBatiment, idSalle) VALUES (?, ?)";
            statement = this.cnx.prepareStatement(query);
            for (Salle salle : batiment.getSalles()) {
                if (idSallesNonAttribue.contains(salle.getId())) {
                    statement.setLong(1, batiment.getId());
                    statement.setLong(2, salle.getId());
                    statement.executeUpdate();
                } else if (idSallesBatiment.contains(salle.getId())) {
                        idSallesBatiment.remove(salle.getId());
                }
            }
            query = "DELETE FROM attribution WHERE idSalle = ?";
            statement = this.cnx.prepareStatement(query);
            for (long idSalleBatiment : idSallesBatiment) {
                statement.setLong(1, idSalleBatiment);
                statement.executeUpdate();
            }
        }
        this.close();
        return execute;
    }

    @Override
    public List<Batiment> getAll(int debut, int parPage) throws Exception {
        cnx = this.sql.connect();
        List<Batiment> batiments = new ArrayList<Batiment>();
        String query = "SELECT * FROM batiment ORDER BY nomBatiment ASC LIMIT ?,?";
        statement = this.cnx.prepareStatement(query);
        statement.setInt(1, debut);
        statement.setInt(2, parPage);
        result = statement.executeQuery();
        while (result.next()) {
            batiments.add(createBatiment(result));
        }
        this.close();
        return batiments;
    }

    @Override
    public int getCount() throws Exception {
        cnx = this.sql.connect();
        int count = 0;
        String query = "SELECT COUNT(*) FROM batiment";
        statement = this.cnx.prepareStatement(query);
        result = statement.executeQuery();
        if (result.next()) {
            count = result.getInt(1);
        }
        this.close();
        return count;
    }

    @Override
    public Batiment getById(long id) throws Exception {
        cnx = this.sql.connect();
        Batiment batiment = null;
        String query = "SELECT * FROM batiment WHERE id = ?";
        statement = this.cnx.prepareStatement(query);
        statement.setLong(1, id);
        result = statement.executeQuery();
        if (result.next()) {
            batiment = createBatiment(result);
        }
        this.close();
        return batiment;
    }

    @Override
    public Batiment getByNomBatiment(String nomBatiment) throws Exception {
        cnx = this.sql.connect();
        Batiment batiment = null;
        String query = "SELECT * FROM batiment WHERE nomBatiment = ?";
        statement = this.cnx.prepareStatement(query);
        statement.setString(1, nomBatiment);
        result = statement.executeQuery();
        if (result.next()) {
            batiment = createBatiment(result);
        }
        this.close();
        return batiment;
    }

    @Override
    public List<String> getListNomBatiment(int debut, int parPage) throws Exception {
        cnx = this.sql.connect();
        List<String> listNomBatiment = new ArrayList<String>();
        String query = "SELECT nomBatiment FROM batiment WHERE nomBatiment != '' ORDER BY nomBatiment ASC LIMIT ?,?";
        statement = this.cnx.prepareStatement(query);
        statement.setInt(1, debut);
        statement.setInt(2, parPage);
        result = statement.executeQuery();
        while (result.next()) {
            listNomBatiment.add(result.getString("nomBatiment"));
        }
        this.close();
        return listNomBatiment;
    }

    @Override
    public int getCountListNomBatiment() throws Exception {
        cnx = this.sql.connect();
        int count = 0;
        String query = "SELECT COUNT(nomBatiment) FROM batiment WHERE idBadge != ''";
        statement = this.cnx.prepareStatement(query);
        result = statement.executeQuery();
        if (result.next()) {
            count = result.getInt(1);
        }
        this.close();
        return count;
    }

    @Override
    public Batiment getBySalle(Salle salle) throws Exception {
        cnx = this.sql.connect();
        Batiment batiment = null;
        String query = "SELECT * FROM batiment WHERE id IN (SELECT idBatiment FROM attribution WHERE idSalle = ?)";
        statement = this.cnx.prepareStatement(query);
        statement.setLong(1, salle.getId());
        result = statement.executeQuery();
        if (result.next()) {
            batiment = createBatiment(result);
        }
        this.close();
        return batiment;
    }

    @Override
    public List<Salle> getListSalle(int debut, int parPage) throws Exception {
        cnx = this.sql.connect();
        List<Salle> salles = new ArrayList<Salle>();
        String query = "SELECT * FROM salle WHERE id IN (SELECT idSalle FROM attribution) ORDER BY numSalle ASC LIMIT ?,?";
        statement = this.cnx.prepareStatement(query);
        statement.setInt(1, debut);
        statement.setInt(2, parPage);
        result = statement.executeQuery();
        while (result.next()) {
            long id = result.getLong("id");
            float x_ini = result.getFloat("x_ini");
            float y_ini = result.getFloat("y_ini");
            float x_fin = result.getFloat("x_fin");
            float y_fin = result.getFloat("y_fin");
            int numSalle = result.getInt("numSalle");
            String nomSalle = result.getString("nomSalle");
            String nomAtelier = result.getString("nomAtelier");
            Salle salle = new Salle(x_ini, y_ini, x_fin, y_fin, numSalle, nomSalle, nomAtelier);
            salle.setId(id);
            salles.add(salle);
        }
        this.close();
        return salles;
    }

    @Override
    public int getCountListSalle() throws Exception {
        cnx = this.sql.connect();
        int count = 0;
        String query = "SELECT COUNT(*) FROM salle WHERE id IN (SELECT idSalle FROM attribution)";
        statement = this.cnx.prepareStatement(query);
        result = statement.executeQuery();
        if (result.next()) {
            count = result.getInt(1);
        }
        this.close();
        return count;
    }

    @Override
    public Batiment getByPlan(String plan) throws Exception {
        cnx = this.sql.connect();
        Batiment batiment = null;
        String query = "SELECT * FROM batiment WHERE plan = ?";
        statement = this.cnx.prepareStatement(query);
        statement.setString(1, plan);
        result = statement.executeQuery();
        if (result.next()) {
            batiment = createBatiment(result);
        }
        this.close();
        return batiment;
    }

    @Override
    public List<String> getListPlan(int debut, int parPage) throws Exception {
        cnx = this.sql.connect();
        List<String> listPlan = new ArrayList<String>();
        String query = "SELECT plan FROM batiment WHERE plan != '' ORDER BY plan ASC LIMIT ?,?";
        statement = this.cnx.prepareStatement(query);
        statement.setInt(1, debut);
        statement.setInt(2, parPage);
        result = statement.executeQuery();
        while (result.next()) {
            listPlan.add(result.getString("plan"));
        }
        this.close();
        return listPlan;
    }

    @Override
    public int getCountListPlan() throws Exception {
        cnx = this.sql.connect();
        int count = 0;
        String query = "SELECT COUNT(plan) FROM batiment WHERE plan != ''";
        statement = this.cnx.prepareStatement(query);
        result = statement.executeQuery();
        if (result.next()) {
            count = result.getInt(1);
        }
        this.close();
        return count;
    }

}

File: AppMateriel/physique/batimentdataserviceimpl.cpp
#include "batimentdataserviceimpl.h"

Physique::BatimentDataServiceImpl::BatimentDataServiceImpl()
{
}

Batiment Physique::BatimentDataServiceImpl::add(Batiment batiment)
{
    Batiment newBatiment;
    long id = 0;
    QList<long> idSallesNonAttribue;
    bool execute;
    query.prepare("INSERT INTO batiment (nomBatiment, plan) VALUES (:nomBatiment, :plan)"); // Requête préparée, les paramètres sont représenté comme ceci: :param
    query.bindValue(":nomBatiment", batiment.getNomBatiment()); // Ajoute les valeurs
    query.bindValue(":plan", batiment.getPlan());
    if(execute = query.exec()) { // Si la requête s'est éxécuté, on réccupère le dernier id inséré dans la bdd
        id = query.lastInsertId().toLongLong(); // Les résultats retournés sont des QVariant, type très puissant permettant de stocker n'importe quel type de variable
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    if(execute){
        query.prepare("SELECT id FROM salle WHERE id NOT IN (SELECT idSalle FROM attribution)"); // séléctionne tous les id des salles qui n'ont pas été attribués à un batiment
        if(query.exec()) {
            while(query.next()){
                idSallesNonAttribue.append(query.value(0).toLongLong());
            }
        }else{
            qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
        }
        query.prepare("INSERT INTO attribution (id, idBatiment, idSalle) VALUES ('', :idBatiment, :idSalle)");
        foreach(Salle salle, batiment.getSalles()){ // Pour chaque salle du batiment
            if(idSallesNonAttribue.contains(salle.getId())){ // si l'id n'est pas attribué, on l'ajoute au batiment
                query.bindValue(":idBatiment", (qlonglong)id);
                query.bindValue(":idSalle", (qlonglong)salle.getId());
                if(!query.exec()) {
                    qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
                }
            }else{
                qWarning("Salle déjà attribuée");
            }
        }
    }
    if(id != 0){ // Si un id a été créé par la bdd, on réccupère le batiment
        newBatiment = getById(id);
    }
    return newBatiment;
}

bool Physique::BatimentDataServiceImpl::remove(Batiment batiment)
{
    bool execute;
    query.prepare("DELETE FROM attribution WHERE idBatiment = :idBatiment"); // Supprime d'abord les attributions qui contient le batiment puis le batiment
    query.bindValue(":idBatiment", (qlonglong)batiment.getId());
    if(execute = query.exec()) {
        qWarning("Attributions supprimées\n");
        query.prepare("DELETE FROM batiment WHERE id = :id");
        query.bindValue(":id", (qlonglong)batiment.getId());
        if(!(execute = query.exec())) {
            qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return execute;
}

bool Physique::BatimentDataServiceImpl::update(Batiment batiment)
{
    QList<long> idSallesBatiment;
    QList<long> idSallesNonAttribue;
    bool execute;
    query.prepare("UPDATE batiment SET nomBatiment = :nomBatiment, plan = :plan WHERE id = :id");
    query.bindValue(":id", (qlonglong)batiment.getId());
    query.bindValue(":nomBatiment", batiment.getNomBatiment());
    query.bindValue(":plan", batiment.getPlan());
    if(!(execute = query.exec())) {
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    if(execute){
        query.prepare("SELECT idSalle FROM attribution WHERE idBatiment = :idBatiment"); // séléctionne tous les id des salles qui ont été attribués à notre batiment
        query.bindValue(":idBatiment", (qlonglong)batiment.getId());
        if(query.exec()) {
            while(query.next()){
                idSallesBatiment.append(query.value(0).toLongLong());
            }
        }else{
            qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
        }
        query.prepare("SELECT id FROM salle WHERE id NOT IN (SELECT idSalle FROM attribution)"); // séléctionne tous les id des salles qui n'ont pas été attribués à un batiment
        if(query.exec()) {
            while(query.next()){
                idSallesNonAttribue.append(query.value(0).toLongLong());
            }
        }else{
            qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
        }
        query.prepare("INSERT INTO attribution (id, idBatiment, idSalle) VALUES ('', :idBatiment, :idSalle)");
        foreach(Salle salle, batiment.getSalles()){
            if(idSallesNonAttribue.contains(salle.getId())){ // Si la salle n'est pas attribué, on l'ajoute
                query.bindValue(":idBatiment", (qlonglong)batiment.getId());
                query.bindValue(":idSalle", (qlonglong)salle.getId());
                if(!query.exec()) {
                    qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
                }
            }else{ // Sinon la salle est déjà attribué, mais l'est-elle à notre batiment?
                if(idSallesBatiment.contains(salle.getId())){
                    idSallesBatiment.removeOne(salle.getId()); // Si oui on la supprime de la liste des anciennes salle attribué à notre batiment
                }else{
                    qWarning("Salle déjà attribuée à un autre batiment\n");
                }
            }
        }
        query.prepare("DELETE FROM attribution WHERE idSalle = :idSalle");
        foreach(long idSalleBatiment, idSallesBatiment){ // Toutes les salles présente dans la liste des anciennes salle attribué à notre batiment ne le sont plus, on les supprime donc
            query.bindValue(":idSalle", (qlonglong)idSalleBatiment);
            if(!query.exec()) {
                qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
            }
        }
    }
    return execute;
}

QList<Batiment> Physique::BatimentDataServiceImpl::getAll(int debut, int parPage)
{
    bool execute;
    QList<Batiment> batiments_temp;
    QList<Batiment> batiments;
    Batiment batiment_temp;
    QList<Salle> salles;
    Salle salle;
    query.prepare("SELECT * FROM batiment ORDER BY nomBatiment ASC LIMIT :debut,:parPage");
    query.bindValue(":debut", debut);
    query.bindValue(":parPage", parPage);
    if(execute = query.exec()) {
        while(query.next()) // Réccupère les infos excepté les salles puisque les infos se trouve dans la table attribution
        {
            batiment_temp = Batiment();
            batiment_temp.setId(query.value(0).toLongLong());
            batiment_temp.setNomBatiment(query.value(1).toString());
            batiment_temp.setPlan(query.value(2).toString());
            batiments_temp.append(batiment_temp);
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    if(execute){
        query.prepare("SELECT * FROM salle WHERE id IN (SELECT idSalle FROM attribution WHERE idBatiment = :idBatiment) ORDER BY numSalle"); // on séléctionne tous les id des salles qui sont attribués au batiment et on les ajoutes
        foreach(Batiment batiment, batiments_temp){
            query.bindValue(":idBatiment", (qlonglong)batiment.getId());
            if(query.exec()) {
                while(query.next())
                {
                    salle = Salle(query.value(1).toInt(), query.value(2).toInt(), query.value(3).toInt(), query.value(4).toInt(), query.value(5).toInt(), query.value(6).toString(), query.value(7).toString());
                    salle.setId(query.value(0).toLongLong());
                    salles.append(salle);
                }
                batiment.setSalles(salles);
                batiments.append(batiment);
            }else{
                qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
            }
            salles.clear(); // A chaque batiment, on vide la liste de salle
        }
    }
    return batiments;
}

int Physique::BatimentDataServiceImpl::getCount()
{
    int count = 0;
    query.prepare("SELECT COUNT(*) FROM batiment"); // Réccupère le nombre total de batiment
    if(query.exec()) {
        if(query.next()){
            count = query.value(0).toInt();
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return count;
}

Batiment Physique::BatimentDataServiceImpl::getById(long id)
{
    bool execute;
    Batiment batiment;
    QList<Salle> salles;
    Salle salle;
    query.prepare("SELECT * FROM batiment WHERE id = :id"); // Recherche par attribut
    query.bindValue(":id", (qlonglong)id);
    if(execute = query.exec()) {
        if(query.next())
        {
            batiment = Batiment();
            batiment.setId(query.value(0).toLongLong());
            batiment.setNomBatiment(query.value(1).toString());
            batiment.setPlan(query.value(2).toString());
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    if(execute){
        query.prepare("SELECT * FROM salle WHERE id IN (SELECT idSalle FROM attribution WHERE idBatiment = :idBatiment) ORDER BY numSalle");
        query.bindValue(":idBatiment", (qlonglong)batiment.getId());
        if(query.exec()) {
            while(query.next())
            {
                salle = Salle(query.value(1).toInt(), query.value(2).toInt(), query.value(3).toInt(), query.value(4).toInt(), query.value(5).toInt(), query.value(6).toString(), query.value(7).toString());
                salle.setId(query.value(0).toLongLong());
                salles.append(salle);
            }
            batiment.setSalles(salles);
        }else{
            qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
        }
    }
    return batiment;
}

Batiment Physique::BatimentDataServiceImpl::getByNomBatiment(QString nomBatiment)
{
    bool execute;
    Batiment batiment;
    QList<Salle> salles;
    Salle salle;
    query.prepare("SELECT * FROM batiment WHERE nomBatiment = :nomBatiment");
    query.bindValue(":nomBatiment", nomBatiment);
    if(execute = query.exec()) {
        if(query.next())
        {
            batiment = Batiment();
            batiment.setId(query.value(0).toLongLong());
            batiment.setNomBatiment(query.value(1).toString());
            batiment.setPlan(query.value(2).toString());
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    if(execute){
        query.prepare("SELECT * FROM salle WHERE id IN (SELECT idSalle FROM attribution WHERE idBatiment = :idBatiment) ORDER BY numSalle");
        query.bindValue(":idBatiment", (qlonglong)batiment.getId());
        if(query.exec()) {
            while(query.next())
            {
                salle = Salle(query.value(1).toInt(), query.value(2).toInt(), query.value(3).toInt(), query.value(4).toInt(), query.value(5).toInt(), query.value(6).toString(), query.value(7).toString());
                salle.setId(query.value(0).toLongLong());
                salles.append(salle);
            }
            batiment.setSalles(salles);
        }else{
            qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
        }
    }
    return batiment;
}

QList<QString> Physique::BatimentDataServiceImpl::getListNomBatiment(int debut, int parPage)
{
    QList<QString> listNomBatiment;
    query.prepare("SELECT nomBatiment FROM batiment WHERE nombatiment != '' ORDER BY nomBatiment ASC LIMIT :debut,:parPage"); // Réccupère toutes les valeurs non vide enregistrées dans la bdd correspondant à la collone nomBatiment
    query.bindValue(":debut", debut);
    query.bindValue(":parPage", parPage);
    if(query.exec()) {
        while(query.next()){
            listNomBatiment.append(query.value(0).toString());
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return listNomBatiment;
}

int Physique::BatimentDataServiceImpl::getCountListNomBatiment()
{
    int count = 0;
    query.prepare("SELECT COUNT(nomBatiment) FROM batiment WHERE nombatiment != ''"); // Réccupère le nombre de données non vide de la collone nomBatiment enregistré dans la bdd
    if(query.exec()) {
        if(query.next()){
            count = query.value(0).toInt();
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return count;
}

Batiment Physique::BatimentDataServiceImpl::getBySalle(Salle salle)
{
    bool execute;
    QList<Salle> salles;
    Salle salle_temp;
    Batiment batiment;
    query.prepare("SELECT * FROM batiment WHERE id IN (SELECT idBatiment FROM attribution WHERE idSalle = :idSalle)");
    query.bindValue(":idSalle", (qlonglong)salle.getId());
    if(execute = query.exec()) {
        if(query.next())
        {
            batiment = Batiment();
            batiment.setId(query.value(0).toLongLong());
            batiment.setNomBatiment(query.value(1).toString());
            batiment.setPlan(query.value(2).toString());
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    if(execute){
        query.prepare("SELECT * FROM salle WHERE id IN (SELECT idSalle FROM attribution WHERE idBatiment = :idBatiment) ORDER BY numSalle");
        query.bindValue(":idBatiment", (qlonglong)batiment.getId());
        if(query.exec()) {
            while(query.next())
            {
                salle_temp = Salle(query.value(1).toInt(), query.value(2).toInt(), query.value(3).toInt(), query.value(4).toInt(), query.value(5).toInt(), query.value(6).toString(), query.value(7).toString());
                salle_temp.setId(query.value(0).toLongLong());
                salles.append(salle_temp);
            }
            batiment.setSalles(salles);
        }else{
            qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
        }
    }
    return batiment;
}

QList<Salle> Physique::BatimentDataServiceImpl::getListSalle(int debut, int parPage)
{
    QList<Salle> salles;
    Salle salle;
    query.prepare("SELECT * FROM salle WHERE id IN (SELECT idSalle FROM attribution) ORDER BY numSalle ASC LIMIT :debut,:parPage");
    query.bindValue(":debut", debut);
    query.bindValue(":parPage", parPage);
    if(query.exec()) {
        while(query.next()){
            salle = Salle(query.value(1).toInt(), query.value(2).toInt(), query.value(3).toInt(), query.value(4).toInt(), query.value(5).toInt(), query.value(6).toString(), query.value(7).toString());
            salle.setId(query.value(0).toLongLong());
            salles.append(salle);
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return salles;
}

int Physique::BatimentDataServiceImpl::getCountListSalle()
{
    int count = 0;
    query.prepare("SELECT COUNT(*) FROM salle WHERE id IN (SELECT idSalle FROM attribution)");
    if(query.exec()) {
        if(query.next()){
            count = query.value(0).toInt();
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return count;
}

Batiment Physique::BatimentDataServiceImpl::getByPlan(QString plan)
{
    bool execute;
    Batiment batiment;
    QList<Salle> salles;
    Salle salle;
    query.prepare("SELECT * FROM batiment WHERE plan = :plan");
    query.bindValue(":plan", plan);
    if(execute = query.exec()) {
        if(query.next())
        {
            batiment = Batiment();
            batiment.setId(query.value(0).toLongLong());
            batiment.setNomBatiment(query.value(1).toString());
            batiment.setPlan(query.value(2).toString());
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    if(execute){
        query.prepare("SELECT * FROM salle WHERE id IN (SELECT idSalle FROM attribution WHERE idBatiment = :idBatiment) ORDER BY numSalle");
        query.bindValue(":idBatiment", (qlonglong)batiment.getId());
        if(query.exec()) {
            while(query.next())
            {
                salle = Salle(query.value(1).toInt(), query.value(2).toInt(), query.value(3).toInt(), query.value(4).toInt(), query.value(5).toInt(), query.value(6).toString(), query.value(7).toString());
                salle.setId(query.value(0).toLongLong());
                salles.append(salle);
            }
            batiment.setSalles(salles);
        }else{
            qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
        }
    }
    return batiment;
}

QList<QString> Physique::BatimentDataServiceImpl::getListPlan(int debut, int parPage)
{
    QList<QString> listPlan;
    query.prepare("SELECT plan FROM batiment WHERE plan != '' ORDER BY plan ASC LIMIT :debut,:parPage");
    query.bindValue(":debut", debut);
    query.bindValue(":parPage", parPage);
    if(query.exec()) {
        while(query.next()){
            listPlan.append(query.value(0).toString());
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return listPlan;
}

int Physique::BatimentDataServiceImpl::getCountListPlan()
{
    int count = 0;
    query.prepare("SELECT COUNT(plan) FROM batiment WHERE plan != ''");
    if(query.exec()) {
        if(query.next()){
            count = query.value(0).toInt();
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return count;
}
